### PR TITLE
Switch to Rettiwt API

### PR DESCRIPTION
**Pull request by Jeto143**
Searching for tweets hasn't been possible in almost a year with our current API tier. This PR implements https://github.com/Rishikant181/Rettiwt-API, a package that lets users circumvent these limitations.

Several parts of the code have been left unchanged (despite them looking a bit unclean/obsolete) to keep the diff to the point.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,14 +5,13 @@
   "main": "5m5v-bot.js",
   "dependencies": {
     "js-yaml": "^3.13.1",
-    "twit": "^2.2.11"
+    "rettiwt-api": "^2.4.2"
   },
   "devDependencies": {
-    "mocha": "^6.2.0"
+    "nodeunit": "^0.11.3"
```
Comment (Jeto143) [inline package.json:11]: Switched back to `nodeunit` (despite it being deprecated) because all our tests were set up for it and I didn't have time to update them (also it's better to do it in a separate PR anyway).

File: 5m5v-config.yaml.example [hunk 2]
```diff
@@ -1,7 +1,3 @@
-consumer_key: Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0A
-consumer_secret: Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa
 users:
-    - access_token: 0000000000000000000-Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0
-      access_token_secret: Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0
-      filters:
-          - english
+  - api_key: Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0Aa0
+    language: english
```
Comment (Jeto143) [inline 5m5v-config.yaml.example:3]: Updated the config to only handle one language per user, for simplicity (we don't need several).

File: test/test-english.js [hunk 3]
```diff
@@ -49,8 +49,6 @@ var matches = [
 
 var falsePositives = [
     "I do not want to go vegan",
-    "I should be a vegan", // this phrasing more than likely isn't about going vegan long-term (e.g. I should be a vegan for Halloween)
-    "Don't tell me I should go vegan",
```
Comment (Jeto143) [inline test/test-english.js:53]: These didn't work (they do match). Not a huge deal imho.

File: lib/twitter.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        this.tweetFilter = tweetFilter
+    }
+
+    async searchAndRetweet({
+        startDate = new Date(),
+        searchInterval = 30 * 1000,
+        retweetDelay = 120 * 1000,
+        isDryRun = false,
+    } = {}) {
+        const rettiwt = new Rettiwt({ apiKey: this.users[0].api_key })
```
Comment (stephan281094) [inline lib/twitter.js]: Don't we want to do this in the constructor?

File: lib/twitter.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    async retweet(tweetId, language) {
+        const apiKey = this.users.find(user => user.language === language)?.api_key ?? null
+
+        if (apiKey === null) {
+            throw Error(`No user found for ${language} language`)
+        }
+
+        const rettiwt = new Rettiwt({ apiKey })
+
+        return await rettiwt.tweet.retweet('abcd')
```
Comment (stephan281094) [inline lib/twitter.js]:
```suggestion
        return rettiwt.tweet.retweet('abcd')
```

File: 5m5v-bot.js [hunk 6]
```diff
@@ -1,140 +1,27 @@
...
-            if(filterName in filterUsers)
-                filterUsers[filterName].push(i)
-            else
-                filterUsers[filterName] = [i]
-        }
+const languages = config.users.map(user => user.language)
 
-    // use first user to stream with
-    var T = users[0]
+const twitter = new Twitter(config.users, new TweetFilter(config.exclude, languages))
```
Comment (stephan281094) [inline 5m5v-bot.js]: Since api keys are per user, wouldn't it make sense to create a separate twitter instance per user as well?
I imagine something like this:
```js
class Twitter {
  constructor(apiKey, language) {
    this.rettiwt = new Rettiwt({ apiKey });
  }

  async searchAndRetweet() {
    await this.rettiwt.tweet.search(…);
  }
}

for (const user of users) {
  const twitter = new Twitter(user.apiKey, user.language);
  twitter.searchAndRetweet();
}
```

**Review by stephan281094 — APPROVED** [review]
Looks good to me. There are probably things that could be improved, but with the codebase being 5 years old, this is great.

File: lib/twitter.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    async retweet(tweetId, language) {
+        const apiKey = this.users.find(user => user.language === language)?.api_key ?? null
+
+        if (apiKey === null) {
+            throw Error(`No user found for ${language} language`)
+        }
+
+        const rettiwt = new Rettiwt({ apiKey })
+
+        return rettiwt.tweet.retweet('abcd')
```
Comment (Jeto143) [inline lib/twitter.js]: @stephan281094 Can you confirm that it made no difference with or without `await` here? I'm still slightly confused about async/await so I'm curious, thanks!

File: lib/twitter.js (hunk 4 above)
Comment (Jeto143) [reply lib/twitter.js]: The reason it's there is because, the way it's currently done, there are two cases: the general one for searching (which can use any of the API keys), and the one for retweeting (which is user-specific). Hence why each function instantiates its own at the moment.

File: 5m5v-bot.js (hunk 6 above)
Comment (Jeto143) [reply 5m5v-bot.js]: This would definitely be nicer from an API's perspective. I thought about it but the way the bot works is by scanning a list of key words (basically "vegan" and its translations) in all languages at the same time, then checking the matches for all languages. If we split it by language, it will mean one search process per language which would lead to more API calls, I think.

Comment (Jeto143) [reply 5m5v-bot.js]: However your comment made me realize a flaw in the current implementation: words are searched using "AND" (default) rather than "OR" (I tested it with just English which has only one word so didn't notice). I'll need to fix this!

File: lib/twitter.js (hunk 4 above)
Comment (stephan281094) [reply lib/twitter.js]: Gotcha 👍

File: lib/twitter.js (hunk 7 above)
Comment (stephan281094) [reply lib/twitter.js]: Yea, correct. `return await` only really makes sense when used in a try/catch block. Jake Archibald has a great article on this: https://jakearchibald.com/2017/await-vs-return-vs-return-await/.

**Comment by Jeto143** [comment]
Hey @stephan281094, since it was taking some time to figure out how to deploy this (see Discord), I've decided to contribute to the `rettiwt-api` repository by implementing a [tweet streaming functionality](https://github.com/Rishikant181/Rettiwt-API/pull/467). It's not yet been released, so I forked it for now to be able to use it here.

I believe it's cleaner this way, but lmk what you think!

**Comment by stephan281094** [comment]
@Jeto143 Nice work! Looks like your change was merged into the `rettiwt` library and has been released in version `2.6.0`, so we could create a pull request to update to this version 🙌
